Avoid duplicate poster generation

swissspidy opened the pull request:
## Summary

Refreshes media modal to ensure it has all the data about poster images etc.

## Relevant Technical Choices

* Refreshes media modal to ensure it has all the data about poster images etc.  
  See https://github.com/google/web-stories-wp/issues/5675#issuecomment-745591608 for explainer
* When inserting a video from the WordPress media modal, do not try to generate a poster image if the video already has a poster.

## To-do

<!-- A list of things that need to be addressed in this PR or follow-up changes. -->

## User-facing changes

* Media modal refreshes when there's newly uploaded media

## Testing Instructions

1. In the WordPress media library (Media -> Add New), upload a video
2. Open dev tools and monitor network tab
2. Create a new story
3. See that video is available in media pane.
4. Optional: Verify that in the network tab you see some XHR requests to `web-stories/v1/media/..` -> a poster is being generated!
5. Click on "Upload" to open media dialog, switch to "Gallery"
6. See that the new video now has a poster, and both the video and the poster are in the list of media items that can be selected (so video + poster = 2 items)
7. Create a new story
8. Click on "Upload", switch to "Gallery"
9. Insert the "video" into page
10. Optional: See that there are _no_ new XHR requests to `web-stories/v1/media/..` -> no extra poster is being generated
11. Click on "Upload" again to open media dialog, switch to "Gallery"
12. See that the video still has a poster, and _no new poster_ has been added. So you still see the video + the poster in the list of selectable items, but not another new one (so video + poster = 2 items)

**Note:** In the WordPress media modal and the WordPress media library, you will see both the video and its poster image, whereas in the story editor the poster is not visible; only the video. This is currently expected.

---

<!-- Please reference the issue(s) this PR addresses. -->

Fixes #5675
Fixes #5541

swissspidy committed:
Remove some dead and incorrect code

Fixes #5541

swissspidy committed:
Use `global.wp`

swissspidy committed:
Force-refresh media modal contents

swissspidy committed:
Lint fix

swissspidy committed:
Fix errors

spacedmonkey commented on assets/src/edit-story/app/media/local/useContextValueProvider.js:
```diff
@@ -156,17 +144,14 @@ export default function useContextValueProvider(reducerState, reducerActions) {
-        }
-      };
-      process();
+      if (
```
Why does change this from async to sync help the issue?

swissspidy replied:
This part does not fix the issue itself, but fixes a code quality issue outlined in #5541

spacedmonkey commented on assets/src/edit-story/components/mediaPicker/useMediaPicker.js:
```diff
@@ -49,7 +49,7 @@ export default function useMediaPicker({
     try {
       // Work around that forces default tab as upload tab.
-      wp.media.controller.Library.prototype.defaults.contentUserSetting = false;
+      global.wp.media.controller.Library.prototype.defaults.contentUserSetting = false;
```
Can you explain why changing this to global helps here or is this a refactor?

swissspidy replied:
Just a code quality refactor to make it clear that this is a global variable. Makes testing easier too.

Note that `global` in the browser will mean `window`, so `window.wp` as expected

spacedmonkey commented on assets/src/edit-story/components/mediaPicker/useMediaPicker.js:
```diff
@@ -91,15 +91,22 @@ export default function useMediaPicker({
 
     // When an image is selected, run a callback.
-    fileFrame.on('select', () => {
+    fileFrame?.once('select', () => {
```
Again, why does once and conditionals help here?

swissspidy replied:
Just a safety measure to avoid inserting elements & generating posters multiple times should the event be fired more than once.

swissspidy committed:
Remove incorrect promises

spacedmonkey commented on assets/src/edit-story/components/mediaPicker/useMediaPicker.js:
```diff
@@ -91,15 +91,22 @@ export default function useMediaPicker({
+    fileFrame?.once('content:activate:browse', () => {
+      // Force-refresh media modal contents every time
+      // to avoid stale data.
+      fileFrame?.content?.get()?.collection?._requery(true);
```
Don't we also need to refresh media library if you drag an image/video into the editor?

swissspidy replied:
Yes, but this code already handles this, because it refreshes when opening the modal.

1. Drag video into editor to upload
2. Open media modal
3. Media modal refreshes automatically, and will include the new video

swissspidy committed:
Add some JSDoc for useMediaPicker

swissspidy committed:
Use useCallback for openMediaPicker

swissspidy committed:
Check for onPermissionError

swissspidy committed:
Remove optional chaining

swissspidy committed:
Fix karma test

swissspidy committed:
Compress e2e images

swissspidy committed:
E2E Assets: add poster for webm video

The mov video doesn’t need a poster because it won’t be allowed in the editor anyway

swissspidy committed:
Merge branch 'main' into fix/refresh-media-modal

spacedmonkey commented on bin/local-env/install-wordpress.sh:
```diff
@@ -88,7 +88,8 @@ container chmod 767 \
 wp media import /var/www/html/wp-content/e2e-assets/small-video.mov
-wp media import /var/www/html/wp-content/e2e-assets/small-video.webm
+WEBM_VIDEO_ID=wp media import /var/www/html/wp-content/e2e-assets/small-video.webm --porcelain
+wp media import /var/www/html/wp-content/e2e-assets/small-video-poster.jpg --post_id=$WEBM_VIDEO_ID
```
```suggestion
POSTER_ID=wp media import /var/www/html/wp-content/e2e-assets/small-video-poster.jpg --post_id=$WEBM_VIDEO_ID --porcelain
wp post meta set $WEBM_VIDEO_ID _thumbnail_id $POSTER_ID
```

swissspidy replied:
Ah! Good call. Can actually use the `--featured_image` arg though, so don't need an extra line.

spacedmonkey replied:
```suggestion
wp media import /var/www/html/wp-content/e2e-assets/small-video-poster.jpg --post_id=$WEBM_VIDEO_ID --featured_image
``` 
So this? 

Didn't know that flag existed. Very cool

swissspidy replied:
Yep :) Just saw it in the docs too.

swissspidy committed:
Actually set poster as featured image

swissspidy committed:
Add some comments

swissspidy committed:
Fix useMediaPicker test

swissspidy committed:
E2E Tests: Fix bash command

swissspidy committed:
Fix e2e test assertion

swissspidy committed:
E2E Tests: Set correct terms and meta data for poster

swissspidy committed:
Fix `wp post meta` ccommand

swissspidy committed:
Merge branch 'main' into fix/refresh-media-modal

swissspidy committed:
Move comment to right place

swissspidy committed:
Make output quiet

swissspidy committed:
Merge branch 'main' into fix/refresh-media-modal

swissspidy committed:
Merge branch 'main' into fix/refresh-media-modal

spacedmonkey committed:
Merge branch 'main' into fix/refresh-media-modal

swissspidy committed:
Merge branch 'main' into fix/refresh-media-modal

swissspidy committed:
Media Pane: Avoid uploading poster if one already exists

Previously, when inserting a video from the WordPress media modal, poster generation was attempted even if the video already has a poster, effectively overwriting it.

swissspidy committed:
Lint fix

spacedmonkey reviewed (approved):
In my testing, this seemed to work extremely well. The poster appears almost straight away now. Nice to see the test work better well. Happy this finally seems to be fixed. 👍

swissspidy committed:
Merge branch 'main' into fix/refresh-media-modal

swissspidy committed:
Merge branch 'main' into fix/refresh-media-modal

swissspidy committed:
Merge branch 'main' into fix/refresh-media-modal
